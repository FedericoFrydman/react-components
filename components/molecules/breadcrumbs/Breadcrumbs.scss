@import '../../assets/styles/colors.scss';

.lsux-crumbs {
  align-items: center;
  display: flex;

  & &__crumb {
    align-items: center;
    display: flex;
  }

  & &__crumb--ellipsis {
    margin: 2px;
    padding: 3px;
  }

  & &__crumb--link {
    color: $B200;
    margin: 2px;
    padding: 3px;
    text-decoration: none;
  }

  & &__crumb--leaf {
    margin: 2px;
    padding: 3px;
  }

  & &__crumb--link:hover {
    text-decoration: underline;
  }

  & &__dropdown-items--wrapper {
    background-color: white;
    border: 1px solid $N300;
    border-radius: var(--brand-border-radius);
    z-index: 1;
  }
}
