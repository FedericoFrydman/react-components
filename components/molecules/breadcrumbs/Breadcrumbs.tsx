import { forwardRef, HTMLProps, Ref, useState } from 'react';
import { usePopper } from 'react-popper';
import cn from 'classnames';

import './Breadcrumbs.scss';
import { Icon } from '../../atoms/icon/Icon';
import { useOutsideClick } from '../../foundation/utilities';

export interface Crumb {
  label: string;
  link?: string;
  pplsiEventId?: string;
}

export interface BreadcrumbsProps extends Omit<HTMLProps<HTMLDivElement>, 'css' | 'onClick'> {
  /**
   * An array of labels and link data representing each item in the crumbs
   */
  crumbs: Crumb[];
  /**
   * Maximum number of crumbs displayed at the tail.  If the crumbs length is greater than 2 more than this
   * an ellipses will be shown between the first item and the tail.
   */
  maxTail: number;
  /**
   * Callback function when a crumb is clicked. Data passed is the link data
   * associated with the crumb
   */
  onClick: (link: string) => void;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLDivElement>;
}

const BreadcrumbsComponent = (
  { crumbs, maxTail = 0, onClick, classNames = [], ...props }: BreadcrumbsProps,
  ref: BreadcrumbsProps['ref'],
) => {
  // The last item in the crumbs should not have a link; this is nulled for sanity
  crumbs[crumbs.length - 1].link = undefined;

  // State representing if there is an open link menu
  const [menuOpen, setMenuOpen] = useState(false);

  // Using popper to handle the menu
  const [referenceElement, setReferenceElement] = useState(null);
  const [popperElement, setPopperElement] = useState(null);
  const { styles, attributes } = usePopper(referenceElement, popperElement, {
    modifiers: [
      {
        name: 'offset',
        options: {
          offset: [0, 8],
        },
      },
    ],
  });

  // When the ellipses is clicked, the menu is opened
  const ellipsesClicked = (e: React.MouseEvent<HTMLAnchorElement>): void => {
    e.preventDefault();
    setMenuOpen(!menuOpen);
  };

  // Call back for a click outside of the menu
  useOutsideClick(
    popperElement,
    () => {
      if (menuOpen) setMenuOpen(false);
    },
    referenceElement,
  );

  const classCrumbs = 'lsux-crumbs';

  // Function that renders a breadcrumb link
  const crumblink = (crumb: Crumb, index?: number, array: Array<Crumb> = []) => {
    const labelElement = crumb.link ? (
      <a
        className={`${classCrumbs}__crumb--link`}
        data-pplsi-event-id={crumb.pplsiEventId}
        href="#"
        onClick={(e) => {
          e.preventDefault();
          onClick(crumb.link);
        }}
      >
        {crumb.label}
      </a>
    ) : (
      <span className={`${classCrumbs}__crumb--leaf`}>{crumb.label}</span>
    );
    return (
      <div key={crumb.label} className={`${classCrumbs}__crumb`}>
        {labelElement}
        {index !== array.length - 1 && <Icon name="arrow_chevron_right" size="small" />}
      </div>
    );
  };

  // Function that renders a bread crumb link in an open menu
  const menulink = (crumb: Crumb) => {
    return (
      <div key={crumb.label} style={{ padding: '8px' }}>
        <a
          className={`${classCrumbs}__crumb--link`}
          data-pplsi-event-id={crumb.pplsiEventId}
          href="#"
          onClick={(e) => {
            e.preventDefault();
            onClick(crumb.link);
          }}
        >
          {crumb.label}
        </a>
      </div>
    );
  };

  if (maxTail != 0 && crumbs.length > maxTail + 2) {
    // Render breadcrumbs with an ellipses and menu
    const first = crumbs[0];
    const mid = crumbs.slice(1, crumbs.length - maxTail);
    const tail = crumbs.slice(crumbs.length - maxTail, crumbs.length);
    return (
      <div {...props} className={cn(`${classCrumbs}`, classNames)} ref={ref}>
        {crumblink(first)}
        <span>
          <a href="#" className={`${classCrumbs}__crumb--link`} ref={setReferenceElement} onClick={ellipsesClicked}>
            &#8230;
          </a>
          {menuOpen && (
            <div
              ref={setPopperElement}
              style={styles.popper}
              className={cn(`${classCrumbs}__dropdown-items--wrapper`, 'shadow300-bottom')}
              {...attributes.popper}
            >
              {mid.map(menulink)}
            </div>
          )}
        </span>
        <Icon name="arrow_chevron_right" size="small" />
        {tail.map(crumblink)}
      </div>
    );
  } else {
    // Render all the crumbs; no ellipse needed
    return (
      <div {...props} className={cn(`${classCrumbs}`, classNames)} ref={ref}>
        {crumbs.map(crumblink)}
      </div>
    );
  }
};

/**
 * Breadcrumbs is a form of secondary navigation typically locates below the page title.
 * It is a trail of links that describes the hierarchical relationship between current page and the overall structure.
 * Breadcrumbs help users navigate your product by understanding their path to current location and way to go back.
 */

export const Breadcrumbs = forwardRef<HTMLDivElement, BreadcrumbsProps>(BreadcrumbsComponent);
