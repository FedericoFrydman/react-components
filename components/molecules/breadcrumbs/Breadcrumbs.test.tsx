import { shallow, ShallowWrapper } from 'enzyme';
import { createRef } from 'react';
import { render, screen } from '@testing-library/react';
import cn from 'classnames';

import { Breadcrumbs, BreadcrumbsProps } from './Breadcrumbs';

describe('Breadcrumbs Component', () => {
  let baseClass: string;
  let testId: string;
  let wrapper: ShallowWrapper;
  let pplsiEventId: Array<string>;
  const crumbs = [
    {
      label: 'Home',
      link: '/',
      pplsiEventId: 'test prop 1',
    },
    {
      label: 'Page1',
      link: '/Page1',
      pplsiEventId: 'test prop 2',
    },
    {
      label: 'Page2',
      link: '',
      pplsiEventId: 'test prop 3',
    },
    {
      label: 'Page3',
      link: '/Page2',
      pplsiEventId: 'test prop 4',
    },
    {
      label: 'Page3',
      link: '',
      pplsiEventId: 'test prop 5',
    },
  ];
  const maxTail = 0;
  const onClick = function (link: string): void {
    throw new Error('Function not implemented.');
  };

  beforeEach(() => {
    baseClass = 'lsux-crumbs';
    testId = 'test Breadcrums';
    wrapper = shallow(<Breadcrumbs crumbs={crumbs} maxTail={maxTail} onClick={onClick} />);
    pplsiEventId = ['test prop 1', 'test prop 2', 'test prop 4'];
  });

  test('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the icon the provided classNames', () => {
          const breadcrumb = renderBreadcrumbs({ classNames, crumbs, maxTail, onClick });

          expect(breadcrumb).toHaveClass(cn(classNames), { exact: false });
        });
      });
    });

    describe('maxTail', () => {
      describe('when dropdown should be working', () => {
        it('gives the breadcrumb maxtail 1', () => {
          const breadcrumb = renderBreadcrumbs({ crumbs, maxTail: 1, onClick });
          const maxtailComp = breadcrumb.getElementsByClassName(`${baseClass}__dropdown-items--wrapper`);

          expect(maxtailComp).toBeTruthy();
        });
      });
    });

    describe('onClick', () => {
      describe('when onClick is present', () => {
        const onClickSpy = jest.fn();

        beforeEach(() => jest.clearAllMocks());

        it('adds the clickable class to the breadcrumb', () => {
          const breadcrumb = renderBreadcrumbs({ crumbs, maxTail, onClick: onClickSpy });
          const onClickClass = `${baseClass}__crumb--link`;

          expect(breadcrumb.querySelector('a')).toHaveClass(onClickClass, { exact: false });
        });

        it('calls the onClick handler when pressed', () => {
          const breadcrumb = renderBreadcrumbs({ crumbs, maxTail, onClick: onClickSpy });

          breadcrumb.click();

          expect(onClickSpy).toHaveBeenCalledTimes(0);
        });
      });
    });
  });

  describe('ref', () => {
    const ref = createRef<HTMLDivElement>();

    it('provides a ref to the breadcrumbs', () => {
      expect(renderBreadcrumbs({ crumbs, maxTail, onClick, ref })).toEqual(ref.current);
    });
  });

  describe('unspecified html attributes', () => {
    it('allows pplsi event id data attribute to be passed to the underlying breadcrumb element', () => {
      render(<Breadcrumbs crumbs={crumbs} maxTail={maxTail} onClick={onClick} data-testid={testId} />);

      const breadcrumb = screen.getByTestId(testId);

      const maxtailComp = breadcrumb.getElementsByClassName(`${baseClass}__crumb--link`);

      const maxtailCompKeys = Object.keys(maxtailComp);

      maxtailCompKeys.forEach((element: any) => {
        expect(maxtailComp[element]).toHaveAttribute('data-pplsi-event-id', pplsiEventId[element]);
      });
    });
  });

  const renderBreadcrumbs = (props: BreadcrumbsProps): HTMLDivElement => {
    render(<Breadcrumbs {...props} data-testid={testId} />);
    return screen.getByTestId(testId);
  };
});
