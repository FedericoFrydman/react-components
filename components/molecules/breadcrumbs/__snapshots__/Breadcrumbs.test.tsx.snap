// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs Component should render the component 1`] = `
<div
  className="lsux-crumbs"
>
  <div
    className="lsux-crumbs__crumb"
    key="Home"
  >
    <a
      className="lsux-crumbs__crumb--link"
      data-pplsi-event-id="test prop 1"
      href="#"
      onClick={[Function]}
    >
      Home
    </a>
    <ForwardRef(IconComponent)
      name="arrow_chevron_right"
      size="small"
    />
  </div>
  <div
    className="lsux-crumbs__crumb"
    key="Page1"
  >
    <a
      className="lsux-crumbs__crumb--link"
      data-pplsi-event-id="test prop 2"
      href="#"
      onClick={[Function]}
    >
      Page1
    </a>
    <ForwardRef(IconComponent)
      name="arrow_chevron_right"
      size="small"
    />
  </div>
  <div
    className="lsux-crumbs__crumb"
    key="Page2"
  >
    <span
      className="lsux-crumbs__crumb--leaf"
    >
      Page2
    </span>
    <ForwardRef(IconComponent)
      name="arrow_chevron_right"
      size="small"
    />
  </div>
  <div
    className="lsux-crumbs__crumb"
    key="Page3"
  >
    <a
      className="lsux-crumbs__crumb--link"
      data-pplsi-event-id="test prop 4"
      href="#"
      onClick={[Function]}
    >
      Page3
    </a>
    <ForwardRef(IconComponent)
      name="arrow_chevron_right"
      size="small"
    />
  </div>
  <div
    className="lsux-crumbs__crumb"
    key="Page3"
  >
    <span
      className="lsux-crumbs__crumb--leaf"
    >
      Page3
    </span>
  </div>
</div>
`;
